### PR TITLE
Regex Strategy: string generation by regular expressions

### PR DESCRIPTION
**Pull request by kurofuneparry**
I'm no good at git and wasn't able to get the hypothesis test-suite working. Please be gentle, this is my second PR ever :)

This PR aims to accomplish the same goal as the other regex PR. However, this one is ready to go and keeps more of the internal simplicity of the original Xeger project. The test suite is mostly based on the one in the original python Xeger project. It is passing in python 2.7 and 3.5 on my machine using a testing script similar to the one included in test_regex.py. (Again, couldn't get hypothesis test suite working due to problems with my environment.)

This is a reimplementation of the Xeger project, but using hypothesis strategies at its base and with the small change of allowing a specified maximum number of repeats to exceed the default limit provided.

I'm happy to make changes or improvements as needed.

### CODE REVIEW

File: src/hypothesis/strategies.py [hunk 1]
```diff
@@ -33,6 +33,7 @@
 from hypothesis.utils.conventions import not_set
 from hypothesis.internal.reflection import proxies
 from hypothesis.searchstrategy.reprwrapper import ReprWrapperStrategy
+from hypothesis.searchstrategy.regex import Xeger
```
Comment (alexwlchan) [inline src/hypothesis/strategies.py]: So I suspect this is the line that’s causing the tests some grief. You import from `regex.py` here… which in turn imports from `strategies.py`… which brings us back to this line. The circular import is throwing everything out of whack. I don’t have time to review properly right now – and this is only a guess, the test output is very opaque – but maybe that’s worth a look?

**Comment by Zac-HD** [comment]
Thanks for contributing ❤️ 

We have issue #662 about a regex strategy, and have been leaning towards merging the [hypothesis-regex](https://github.com/maximkulkin/hypothesis-regex) package.  So it's unfortunate that you also started on a regex strategy 😢 - you'll need to decide whether you want to put a lot more work in to integrate everything, or to leave this and (hopefully!) provide feedback on the final regex pull.

**Comment by kurofuneparry** [comment]
@Zac-HD : Sorry to hear about that. I only found the previous PR when I looked around and the IRC channel didn't mention this other project. My PR doesn't incorporate flags and I don't understand the reduction scheme in hypothesis well enough to know if mine will effectively reduce strings to the empty string where possible.

I'm interested in providing feedback, in part because I was working on this regex generator to make a json-schema generator. I've made a working version that needs improvement, and will likely start working on it with hypothesis-regex for the regex portions of json-shema.

I'm new to git (and competent software design). Should I just delete this fork and PR?

Have a nice weekend.

**Comment by Zac-HD** [comment]
I'm sorry you're in this position, but we'd love your feedback on the final pull - especially if you have a use-case and can describe how porting over affects you.

If you don't mind, it would be nice to keep this pull active for a while longer as a reference.

Finally, however new you are to it this *is* competent design 😄.  If you want any feedback on other code, let me know and I'll be happy to review it for you - especially but not exclusively Hypothesis-related code (eg maximkulkin/hypothesis-regex#6) - or have a more general discussion about design or process.

**Comment by Zac-HD** [comment]
I'm closing this as superseded by #776, because we have enough open pulls that it's starting to get confusing.

**Pull request closed without merging**